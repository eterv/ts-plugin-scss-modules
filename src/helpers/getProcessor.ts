import postcss, { AcceptedPlugin } from 'postcss';
import Processor from 'postcss/lib/processor';
import postcssIcssKeyframes from 'postcss-icss-keyframes';
import postcssIcssSelectors from 'postcss-icss-selectors';

export const getProcessor = (
  additionalPlugins: AcceptedPlugin[] = [],
): Processor =>
  postcss([
    ...additionalPlugins,
    postcssIcssKeyframes(),
    postcssIcssSelectors({
      mode: 'local',
    }),
  ]);
