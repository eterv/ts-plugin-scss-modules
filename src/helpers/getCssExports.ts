import path from 'path';
import Processor from 'postcss/lib/processor';
import sass from 'sass';
import { CSSExports, extractICSS } from 'icss-utils';
import { RawSourceMap } from 'source-map-js';
import tsModule from 'typescript/lib/tsserverlibrary';
import { createMatchPath } from 'tsconfig-paths';
import { sassTildeImporter } from '../importers/sassTildeImporter';
import { Options, CustomRenderer } from '../options';
import { Logger } from './logger';
import { pathToFileURL } from 'url';

export const enum FileType {
  css = 'css',
  sass = 'sass',
  scss = 'scss',
}

export const getFileType = (fileName: string): FileType => {
  if (fileName.endsWith('.css')) return FileType.css;
  if (fileName.endsWith('.sass')) return FileType.sass;
  return FileType.scss;
};

const getFilePath = (fileName: string) => path.dirname(fileName);

export interface CSSExportsWithSourceMap {
  classes: CSSExports;
  css?: string;
  sourceMap?: RawSourceMap;
}

export const getCssExports = ({
  css,
  fileName,
  logger,
  options,
  processor,
  compilerOptions,
}: {
  css: string;
  fileName: string;
  logger: Logger;
  options: Options;
  processor: Processor;
  compilerOptions: tsModule.CompilerOptions;
}): CSSExportsWithSourceMap => {
  try {
    const fileType = getFileType(fileName);
    const rendererOptions = options.rendererOptions || {};

    let transformedCss = '';
    let sourceMap: string | undefined;

    if (options.customRenderer) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const customRenderer = require(options.customRenderer) as CustomRenderer;
      transformedCss = customRenderer(css, {
        fileName,
        logger,
        compilerOptions,
      });
    } else if (fileType === FileType.sass || fileType === FileType.scss) {
      const filePath = getFilePath(fileName);
      const { loadPaths, ...sassOptions } = rendererOptions.sass || {};
      const { baseUrl, paths } = compilerOptions;
      const matchPath =
        baseUrl && paths ? createMatchPath(path.resolve(baseUrl), paths) : null;

      const aliasImporter: sass.FileImporter<'sync'> = {
        findFileUrl(url) {
          const newUrl =
            matchPath !== null
              ? matchPath(url, undefined, undefined, [
                  `.${FileType.sass}`,
                  `.${FileType.scss}`,
                ])
              : undefined;
          return newUrl ? pathToFileURL(newUrl) : null;
        },
      };

      const importers = [aliasImporter, sassTildeImporter];

      const result = sass.compile(fileName, {
        loadPaths: [filePath, 'node_modules', ...(loadPaths || [])],
        importers,
        sourceMap: true,
        ...sassOptions,
      });

      transformedCss = result.css.toString();
      sourceMap = JSON.stringify(result.sourceMap);
    } else {
      transformedCss = css;
    }

    const processedCss = processor.process(transformedCss, {
      from: fileName,
      map: {
        inline: false,
        prev: sourceMap,
      },
    });

    return {
      classes: processedCss.root
        ? extractICSS(processedCss.root).icssExports
        : {},
      css: processedCss.css,
      sourceMap: processedCss.map.toJSON(),
    };
  } catch (e: any) {
    logger.error(e);
    return { classes: {} };
  }
};
