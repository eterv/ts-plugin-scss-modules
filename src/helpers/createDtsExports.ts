import { SourceMapConsumer } from 'source-map-js';
import { CustomTemplate, Options } from '../options';
import { transformClasses } from './classTransforms';
import { CSSExportsWithSourceMap } from './getCssExports';
import { VALID_VARIABLE_REGEXP } from './validVarRegexp';
import { Logger } from './logger';

const isValidVariable = (className: string) =>
  VALID_VARIABLE_REGEXP.test(className);

const classNameToProperty = (className: string) => `'${className}': string;`;
const classNameToNamedExport = (className: string) =>
  `export let ${className}: string;`;

const flattenClassNames = (
  previousValue: string[] = [],
  currentValue: string[],
) => previousValue.concat(currentValue);

export const createDtsExports = ({
  cssExports,
  fileName,
  logger,
  options,
}: {
  cssExports: CSSExportsWithSourceMap;
  fileName: string;
  logger: Logger;
  options: Options;
}): string => {
  const classes = cssExports.classes;

  const processedClasses = Object.keys(classes)
    .map(transformClasses(options.classnameTransform))
    .reduce(flattenClassNames, []);
  const filteredClasses = processedClasses
    .filter(isValidVariable)
    .map(classNameToNamedExport);

  let dts = `\
declare let classes: {
${processedClasses.map(classNameToProperty).join('\n  ')}
};
export default classes;
`;

  if (options.namedExports !== false && filteredClasses.length) {
    dts += filteredClasses.join('\n') + '\n';
  }

  if (options.goToDefinition) {
    if (!cssExports.sourceMap) return dts;

    // Create a new source map consumer.
    const sourceMapConsumer = new SourceMapConsumer(cssExports.sourceMap);

    // Split original CSS file into lines.
    const cssLines = cssExports.css?.split('\n') || [];

    // Create new equal size array of empty strings.
    const dtsLines = Array.from(Array(cssLines.length), () => [] as string[]);

    const processedClasses = Object.entries(classes)
      .map(([className, hashedClassName]) =>
        transformClasses(options.classnameTransform)(className).map(
          (className) => [className, hashedClassName],
        ),
      )
      .reduce((prev, curr) => prev.concat(curr), []);

    const filteredClasses = processedClasses.filter(([className]) =>
      isValidVariable(className as string),
    );

    filteredClasses.forEach(([className, hashedClassName]) => {
      const matchedLine = cssLines.findIndex((line) =>
        line.includes(hashedClassName),
      );
      const matchedColumn =
        matchedLine && cssLines[matchedLine].indexOf(hashedClassName);
      const { line: lineNumber } = sourceMapConsumer.originalPositionFor({
        // Lines start at 1, not 0.
        line: matchedLine >= 0 ? matchedLine + 1 : 1,
        column: matchedColumn >= 0 ? matchedColumn : 0,
      });

      const index = lineNumber ? lineNumber - 1 : 0;

      const prev = dtsLines[index];
      dtsLines[index] = prev == null ? [className] : [...prev, className];
    });

    dts = dtsLines
      .map((classNames) => {
        return classNames.reduce((prev, className) => {
          return `${prev}${classNameToNamedExport(className)}`;
        }, '');
      })
      .join('\n');
  }

  if (options.customTemplate) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const customTemplate = require(options.customTemplate) as CustomTemplate;
    return customTemplate(dts, {
      classes,
      fileName,
      logger,
    });
  }

  return dts;
};
