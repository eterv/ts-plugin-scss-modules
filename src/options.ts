import { Options as SassOptions } from 'sass';
import tsModule from 'typescript/lib/tsserverlibrary';
import { DotenvConfigOptions } from 'dotenv/types';
import { CSSExports } from 'icss-utils';
import { Logger } from './helpers/logger';

export interface PostcssOptions {
  excludePlugins?: string[];
  useConfig?: boolean;
}

export interface RendererOptions {
  sass?: Partial<SassOptions<'sync'>>;
}

export interface Options {
  classnameTransform?: ClassnameTransformOptions;
  customMatcher?: string;
  customRenderer?: string;
  customTemplate?: string;
  dotenvOptions?: DotenvConfigOptions;
  goToDefinition?: boolean;
  namedExports?: boolean;
  postcssOptions?: PostcssOptions;
  /** @deprecated To align with other projects. */
  postCssOptions?: PostcssOptions;
  rendererOptions?: RendererOptions;
}

export type ClassnameTransformOptions =
  | 'asIs'
  | 'camelCase'
  | 'camelCaseOnly'
  | 'dashes'
  | 'dashesOnly';

export interface CustomRendererOptions {
  fileName: string;
  logger: Logger;
  compilerOptions: tsModule.CompilerOptions;
}

export type CustomRenderer = (
  css: string,
  options: CustomRendererOptions,
) => string;

export interface CustomTemplateOptions {
  classes: CSSExports;
  fileName: string;
  logger: Logger;
}

export type CustomTemplate = (
  dts: string,
  options: CustomTemplateOptions,
) => string;
