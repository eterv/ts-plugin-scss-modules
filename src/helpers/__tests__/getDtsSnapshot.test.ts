import { readFileSync } from 'fs';
import { join } from 'path';
import postcssImportSync from 'postcss-import-sync2';
import tsModule from 'typescript/lib/tsserverlibrary';
import { CSSExportsWithSourceMap, getCssExports } from '../getCssExports';
import { createDtsExports } from '../createDtsExports';
import { Logger } from '../logger';
import { getProcessor } from '../getProcessor';
import { Options } from '../../options';

const testFileNames = [
  'test.module.css',
  'test.module.scss',
  'test.module.sass',
  'empty.module.sass',
  'empty.module.scss',
  'import.module.css',
];

const logger: Logger = {
  log: jest.fn(),
  error: jest.fn(),
};

const options: Options = {};

const compilerOptions: tsModule.CompilerOptions = {};

const processor = getProcessor([
  // For testing PostCSS import support/functionality.
  postcssImportSync(),
]);

describe('utils / cssSnapshots', () => {
  testFileNames.forEach((testFile) => {
    let cssExports: CSSExportsWithSourceMap;
    const fileName = join(__dirname, 'fixtures', testFile);
    const css = readFileSync(fileName, 'utf8');

    beforeAll(() => {
      cssExports = getCssExports({
        css,
        fileName,
        logger,
        options,
        processor,
        compilerOptions,
      });
    });

    describe(`with file '${testFile}'`, () => {
      describe('getCssExports', () => {
        it('should return an object matching expected CSS', () => {
          expect(getCssExports).toMatchSnapshot();
        });
      });

      describe('createExports', () => {
        it('should create an exports file', () => {
          const dts = createDtsExports({
            cssExports,
            fileName,
            logger,
            options: {},
          });
          expect(dts).toMatchSnapshot();
        });
      });

      describe('with a custom template', () => {
        it('should transform the generated dts', () => {
          const customTemplate = join(
            __dirname,
            'fixtures',
            'customTemplate.js',
          );

          const options: Options = { customTemplate };

          const dts = createDtsExports({
            cssExports,
            fileName,
            logger,
            options,
          });
          expect(dts).toMatchSnapshot();
        });
      });
    });
  });

  describe('with a Bootstrap import', () => {
    const fileName = join(__dirname, 'fixtures', 'bootstrap.module.scss');
    const css = readFileSync(fileName, 'utf8');

    it('should find external files', () => {
      const cssExports = getCssExports({
        css,
        fileName,
        logger,
        options,
        processor,
        compilerOptions,
      });

      expect(cssExports.classes.test).toMatchSnapshot();
    });
  });

  describe('with a custom renderer', () => {
    const fileName = 'exampleFileContents';
    const css = 'exampleFileName';
    const customRenderer = join(__dirname, 'fixtures', 'customRenderer.js');

    const options: Options = { customRenderer };

    it('should process a file and log', () => {
      const cssExports = getCssExports({
        css,
        fileName,
        logger,
        options,
        processor,
        compilerOptions,
      });

      expect(cssExports.classes).toMatchSnapshot();
      expect(logger.log).toHaveBeenCalledWith('Example log');
    });
  });

  describe('with loadPaths in sass options', () => {
    const fileName = join(__dirname, 'fixtures', 'include-path.module.scss');
    const css = readFileSync(fileName, 'utf8');

    const options: Options = {
      rendererOptions: {
        sass: { loadPaths: [join(__dirname, 'external')] },
      },
    };

    it('should find external file from loadPaths', () => {
      const cssExports = getCssExports({
        css,
        fileName,
        logger,
        options,
        processor,
        compilerOptions,
      });

      expect(cssExports.classes).toMatchSnapshot();
    });
  });

  describe('with baseUrl and paths in compilerOptions', () => {
    const fileName = join(__dirname, 'fixtures', 'tsconfig-paths.module.scss');
    const css = readFileSync(fileName, 'utf8');
    const compilerOptions = {
      baseUrl: __dirname,
      paths: {
        '@scss/*': ['external/package/*'],
        'alias.scss': ['external/package/public.module.scss'],
      },
    };

    it('sass should find the files', () => {
      const cssExports = getCssExports({
        css,
        fileName,
        logger,
        options,
        processor,
        compilerOptions,
      });

      expect(cssExports.classes).toMatchSnapshot();
    });
  });

  describe('with goToDefinition enabled', () => {
    const fileName = join(__dirname, 'fixtures', 'test.module.scss');
    const css = readFileSync(fileName, 'utf8');
    const options: Options = {
      classnameTransform: 'camelCaseOnly',
      goToDefinition: true,
    };

    const cssExports = getCssExports({
      css,
      fileName,
      logger,
      options,
      processor,
      compilerOptions,
    });

    it('should return an object with classes, css, and a source map', () => {
      expect(cssExports).toMatchSnapshot();
    });

    it.only('should return a line-accurate dts file', () => {
      const dts = createDtsExports({
        cssExports,
        fileName,
        logger,
        options,
      });
      expect(dts).toMatchSnapshot();
    });
  });
});
