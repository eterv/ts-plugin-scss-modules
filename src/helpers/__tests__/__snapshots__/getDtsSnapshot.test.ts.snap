// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils / cssSnapshots with goToDefinition enabled should return a line-accurate dts file 1`] = `
"declare let classes: {

};
export default classes;
"
`;
